pkg/checkers/tcpping/tcpping.go

```diff
@@ -0,0 +1,167 @@
+package tcpping
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/Azure/kdebug/pkg/base"
+
+	"github.com/fatih/color"
+	"github.com/shirou/gopsutil/process"
+	log "github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const KubernetesServiceHost = "KUBERNETES_SERVICE_HOST"
+const timeOut = 1000 * time.Millisecond
+
+const (
+	GoogleTarget = "google.com:443"
+)
+
+type pingEndpoint struct {
+	ServerAddress string
+	Name          string
+	NameSpace     string
+}
+
+func (t *TCPChecker) ping(serverAddr string) error {
+	conn, err := t.dialer.Dial("tcp", serverAddr)
+	if err != nil {
+		return err
+	}
+	conn.Close()
+	return nil
+}
+
+type TCPChecker struct {
+	dialer  net.Dialer
+	targets []pingEndpoint
+}
+
+func New() *TCPChecker {
+	return &TCPChecker{
+		dialer: net.Dialer{
+			Timeout: timeOut,
+		},
+	}
+}
+
+func (t *TCPChecker) Name() string {
+	return "TcpChecker"
+}
+
+func (t *TCPChecker) Check(ctx *base.CheckContext) ([]*base.CheckResult, error) {
+	var results []*base.CheckResult
+	targets := append(t.targets, getCheckTargets(ctx)...)
+	var result *base.CheckResult
+	for _, pingTarget := range targets {
+		err := t.ping(pingTarget.ServerAddress)
+		if err != nil {
+			result = &base.CheckResult{
+				Checker: t.Name(),
+				Error:   err.Error(),
+				Description: color.RedString(fmt.Sprintf("Fail to establish tcp connection to %s (%s) in namespace %s.",
+					pingTarget.ServerAddress, pingTarget.Name, pingTarget.NameSpace)),
+				Recommendations: []string{"This might be expected for example the firewall blocks the traffic"},
+			}
+		} else {
+			result = &base.CheckResult{
+				Checker:     t.Name(),
+				Description: color.GreenString(fmt.Sprintf("Successfully establish tcp connection to %s (%s) in namespace %s", pingTarget.ServerAddress, pingTarget.Name, pingTarget.NameSpace)),
```

Comment:
It seems not all targets have a namespace?

Reply:
fixed